tidy3d/plugins/smatrix/component_modelers/terminal.py

```diff
@@ -0,0 +1,425 @@
+                for port, center in zip(self.ports, snap_centers)
+            ]
+            port_current_monitors = [
+                self.to_current_monitor(port=port, snap_center=center)
+                for port, center in zip(self.ports, snap_centers)
+            ]
+            lumped_resistors = [
+                self.to_load(port=port, snap_center=center)
+                for port, center in zip(self.ports, snap_centers)
+            ]
+            new_mnts = (
+                list(self.simulation.monitors) + port_voltage_monitors + port_current_monitors
+            )
+            for port, center in zip(self.ports, snap_centers):
+                port_source = self.to_source(port=port, snap_center=center)
+                update_dict = dict(
+                    sources=[port_source],
+                    monitors=new_mnts,
+                    lumped_elements=lumped_resistors,
+                    grid_spec=grid_spec,
+                )
+
+                sim_copy = self.simulation.copy(update=update_dict)
+                task_name = self._task_name(port=port)
+                sim_dict[task_name] = sim_copy
+
+        return sim_dict
+
+    def _construct_smatrix(self, batch_data: BatchData) -> LumpedPortDataArray:
+        """Post process `BatchData` to generate scattering matrix."""
+
+        port_names = [port.name for port in self.ports]
+
+        values = np.zeros(
+            (len(port_names), len(port_names), len(self.freqs)),
+            dtype=complex,
+        )
+        coords = dict(
+            port_out=port_names,
+            port_in=port_names,
+            f=np.array(self.freqs),
+        )
+        a_matrix = LumpedPortDataArray(values, coords=coords)
+        b_matrix = a_matrix.copy(deep=True)
+
+        def port_voltage(port: LumpedPort, sim_data: SimulationData) -> xr.DataArray:
+            """Helper to compute voltage across the port."""
+            e_component = "xyz"[port.voltage_axis]
+            field_data = sim_data[f"{port.name}_E{e_component}"]
+            e_field = field_data.field_components[f"E{e_component}"]
+            e_coords = [e_field.x, e_field.y, e_field.z]
+            # interpolate E and H locations to coincide with port bounds along the integration path
+            e_coords_interp = {
+                e_component: np.linspace(
+                    port.bounds[0][port.voltage_axis],
+                    port.bounds[1][port.voltage_axis],
+                    len(e_coords[port.voltage_axis]),
+                )
+            }
+            e_field = e_field.interp(**e_coords_interp)
+            voltage = -e_field.integrate(coord=e_component).squeeze(drop=True)
+            # Return data array of voltage with coordinates of frequency
+            return voltage
+
+        def port_current(port: LumpedPort, sim_data: SimulationData) -> xr.DataArray:
+            """Helper to compute current flowing through the port."""
```

Comment:
Would be helpful to add more explanation here on how current is computed. `port_voltage` is quite clear, but `port_current` seems quite complicated.

Reply:
I added a small diagram, which should help